### PR TITLE
Gtp refit support

### PR DESCRIPTION
**Pull request by wdykas**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
This is tested and I review but is AI generated to support GTP refit natively.  It also greatly simplifies the planner. The majority of the actual line changes are testing related and naming.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [2386a45](https://github.com/NVIDIA/Megatron-LM/pull/6133/commits/2386a457b7bb5958e0da32c209bea73049dd9747)

**Comment by kvareddy** [comment]
@fanshiqing  can you please take a look at this MR?

File: megatron/core/resharding/README.md [hunk 1]
```diff
@@ -88,22 +89,42 @@ through the network stack.
 
 ## How the Reshard Plan Works
 
-1. Each rank extracts parameter metadata (shape, sharding, TP/EP/PP groups).
+1. Each rank extracts parameter metadata (shape, sharding, TP/GTP/EP/PP groups).
```
Comment (fanshiqing) [inline megatron/core/resharding/README.md]: GTP (generalized tensor parallelism) is a superset of TP, where 

`GTP=TPxGTP_remat`, 

so here should be TP/GTP_remat/EP/PP groups, same changes required below.

See GTP [README](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/api-guide/core/generalized_tensor_parallel.md#generalized-tensor-parallelism-gtp):
> GTP splits the weight-parallel domain into two orthogonal sub-axes — GTP = TP × GTP_remat — so every rank stores 1/(TP × GTP_remat) of each linear weight, together with the matching slice of its gradient and optimizer state.

File: megatron/core/resharding/refit.py [hunk 2]
```diff
@@ -34,16 +34,27 @@
...
+class _ParallelConfig(NamedTuple):
+    """Parallel group sizes that determine a refit plan."""
+
+    tp_size: int
+    pp_size: int
+    ep_size: int
+    dp_size: int
+    expert_tp_size: int
+    gtp_size: int
+    expert_gtp_size: int
```
Comment (fanshiqing) [inline megatron/core/resharding/refit.py]: `gtp_size` -> `gtp_remat_size`
`expert_gtp_size` -> `expert_gtp_remat_size`

File: megatron/core/resharding/execution.py [hunk 3]
```diff
@@ -37,22 +39,52 @@ class _Writeback:
...
-    """Get or lazily allocate the BF16 accumulation buffer for an MXFP8 dest param.
+def _requires_bf16_staging(param: torch.Tensor) -> bool:
+    """Return whether refit must materialize this parameter in BF16.
+
+    Quantized source storage is dequantized before slicing for transfer. On the
+    destination, updating quantized storage slice-by-slice is unsafe because
+    scale blocks can cross slice boundaries, so refit assembles the complete
+    local BF16 weight and quantizes it once after all receives finish.
+    """
+    is_gtp = getattr(param, "is_gtp_weight_remat", False)
```
Comment (fanshiqing) [inline megatron/core/resharding/execution.py]: `getattr(param, "is_gtp_weight_remat", False)` -> gtp_api.is_gtp_param(param)

File: megatron/core/resharding/execution.py [hunk 4]
```diff
@@ -37,22 +39,52 @@ class _Writeback:
...
+        entry = (dst_param, full_bf16)
         pending[param_id] = entry
-    return entry[1], entry[2]
+    return entry[1]
+
+
+def _native_gtp_load_context(module: torch.nn.Module | None, pending: dict[int, tuple]):
+    """Return the special update context required by native quantized GTP params."""
+    if module is None or not any(
+        getattr(param, "is_gtp_weight_remat", False) for param, _buffer in pending.values()
```
Comment (fanshiqing) [inline megatron/core/resharding/execution.py]: `getattr(param, "is_gtp_weight_remat", False)` -> gtp_api.is_gtp_param(param)

File: megatron/core/resharding/transforms.py [hunk 5]
```diff
@@ -105,6 +106,14 @@ def _ensure_sendable(param: torch.Tensor) -> torch.Tensor:
     dequantized to their original precision (usually BF16).  Standard
     parameters are returned via ``.data`` (unwrapped from autograd).
     """
+    if getattr(param, "is_gtp_weight_remat", False) and is_float8tensor(param):
```
Comment (fanshiqing) [inline megatron/core/resharding/transforms.py]: `getattr(param, "is_gtp_weight_remat", False)` -> gtp_api.is_gtp_param(param)

File: megatron/core/resharding/README.md [hunk 6]
```diff
@@ -88,22 +89,42 @@ through the network stack.
...
+     packed/strided TP, GTP, and combined TP x GTP layouts.
    - Assign a monotonic `task_id` per sub-op.  Because the iteration order and
      counter are a pure function of the gathered metadata, the send op computed
      on the sender and the recv op computed on the receiver get the **same**
      `task_id` without any central authority.
 4. Each rank keeps only the ops where it is the sender or receiver.
 5. The plan is cached so repeated refits skip steps 1-4.
 
+## Generalized Tensor Parallelism
+
```
Comment (fanshiqing) [inline megatron/core/resharding/README.md]: Might add a link to GTP userguide inscase some implementation details might be changed later: https://github.com/NVIDIA/Megatron-LM/blob/main/docs/api-guide/core/generalized_tensor_parallel.md#generalized-tensor-parallelism-gtp

File: megatron/core/resharding/execution.py [hunk 7]
```diff
@@ -37,22 +39,52 @@ class _Writeback:
...
+    """Return whether refit must materialize this parameter in BF16.
+
+    Quantized source storage is dequantized before slicing for transfer. On the
+    destination, updating quantized storage slice-by-slice is unsafe because
+    scale blocks can cross slice boundaries, so refit assembles the complete
+    local BF16 weight and quantizes it once after all receives finish.
+    """
+    is_gtp = getattr(param, "is_gtp_weight_remat", False)
+    # MXFP8 staging predates GTP. Broaden it to TE's generic quantized tensor
+    # class only for GTP, adding NVFP4 without changing non-GTP behavior.
```
Comment (fanshiqing) [inline megatron/core/resharding/execution.py]: `adding NVFP4 without changing non-GTP behavior`: sorry I did not get what does it mean here?

File: megatron/core/resharding/execution.py [hunk 8]
```diff
@@ -37,22 +39,52 @@ class _Writeback:
...
+
+
+def _native_gtp_load_context(module: torch.nn.Module | None, pending: dict[int, tuple]):
+    """Return the special update context required by native quantized GTP params."""
+    if module is None or not any(
+        getattr(param, "is_gtp_weight_remat", False) for param, _buffer in pending.values()
+    ):
+        return nullcontext()
+
+    if not gtp_api.HAVE_GTP:
```
Comment (fanshiqing) [inline megatron/core/resharding/execution.py]: check of `gtp_api.HAVE_GTP` should go first (before L79)

Comment (wdykas) [reply megatron/core/resharding/execution.py]: done

File: megatron/core/resharding/execution.py (hunk 7 above)
Comment (wdykas) [reply megatron/core/resharding/execution.py]: In earlier te is_float8tensor checks the generic QuantizedTensor base despite its name, so it also covers native NVFP4 tensors. Before this PR, ordinary refit parameters used BF16 staging only for MXFP8Tensor. The second clause extends staging to other native quantized formats only for GTP parameter which adds NVFP4 support while preserving the existing non-GTP path. I changed the wording if that makes sense? or maybe I am wrong here

File: megatron/core/resharding/transforms.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/resharding/transforms.py]: done

File: megatron/core/resharding/execution.py (hunk 4 above)
Comment (wdykas) [reply megatron/core/resharding/execution.py]: done

File: megatron/core/resharding/execution.py (hunk 3 above)
Comment (wdykas) [reply megatron/core/resharding/execution.py]: done

File: megatron/core/resharding/refit.py (hunk 2 above)
Comment (wdykas) [reply megatron/core/resharding/refit.py]: done

**Review by fanshiqing — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [8d82590](https://github.com/NVIDIA/Megatron-LM/commit/8d82590f80d55c32534cc2e2ba2f4d98c10d0377)

**Comment by wdykas** [comment]
/ok to test [a0fa9c9](https://github.com/NVIDIA/Megatron-LM/pull/6133/commits/a0fa9c98913c92435b75471539eee7d4b8775784)
